app/src/main/java/protect/card_locker/ManageGroupCursorAdapter.java

```diff
@@ -0,0 +1,118 @@
+        LoyaltyCard loyaltyCard = LoyaltyCard.toLoyaltyCard(inputCursor);
+        Boolean overlayValue = mInGroupOverlay.get(loyaltyCard.id);
+        if((overlayValue != null? overlayValue: isLoyaltyCardInGroup(loyaltyCard.id))) {
+            mAnimationItemsIndex.put(inputCursor.getPosition(), true);
+            mSelectedItems.put(inputCursor.getPosition(), true);
+        }
+        mIndexCardMap.put(inputCursor.getPosition(), loyaltyCard.id);
+        super.onBindViewHolder(inputHolder, inputCursor);
+    }
+
+    private List<Group> getGroups(int cardId){
+        List<Group> cache = mGetGroupCache.get(cardId);
+        if(cache != null){
+            return cache;
+        }
+        List<Group> groups = mDb.getLoyaltyCardGroups(cardId);
+        mGetGroupCache.put(cardId, groups);
+        return groups;
+    }
+
+    private boolean isLoyaltyCardInGroup(int cardId){
+        Boolean cache = mIsLoyaltyCardInGroupCache.get(cardId);
+        if(cache != null){
+            return cache;
+        }
+        List<Group> groups = getGroups(cardId);
+        if (groups.contains(mGroup)){
+            mIsLoyaltyCardInGroupCache.put(cardId, true);
+            return true;
+        }
+        mIsLoyaltyCardInGroupCache.put(cardId, false);
+        return false;
+    }
+
+    @Override
+    public void toggleSelection(int inputPosition){
+        super.toggleSelection(inputPosition);
+        Integer cardId = mIndexCardMap.get(inputPosition);
+        if (cardId == null){
+            throw(new RuntimeException("cardId should not be null here"));
+        }
+        Boolean overlayValue = mInGroupOverlay.get(cardId);
+        if (overlayValue == null){
+            mInGroupOverlay.put(cardId, !isLoyaltyCardInGroup(cardId));
+        }else{
+            mInGroupOverlay.remove(cardId);
+        }
+    }
+
+    public boolean hasChanged() {
+        return mInGroupOverlay.size() > 0;
+    }
+
+    public void commitToDatabase(){
+        for(Map.Entry<Integer, Boolean> entry: mInGroupOverlay.entrySet()){
+            int cardId = entry.getKey();
+            List<Group> groups = getGroups(cardId);
+            if(entry.getValue()){
+                groups.add(mGroup);
+            }else{
+                groups.remove(mGroup);
+            }
+            mDb.setLoyaltyCardGroups(cardId, groups);
+        }
+    }
+
+    public void importInGroupState(HashMap<Integer, Boolean> cardIdInGroupMap) {
+        mInGroupOverlay = new HashMap<>();
+        for (Map.Entry<Integer, Boolean> entry: cardIdInGroupMap.entrySet()){
+            mInGroupOverlay.put(entry.getKey(), entry.getValue());
+        }
+    }
+
+    public HashMap<Integer, Boolean> exportInGroupState(){
+        HashMap<Integer, Boolean> ret = new HashMap<>();
+        for (Map.Entry<Integer, Boolean> entry: mInGroupOverlay.entrySet()){
+            ret.put(entry.getKey(), entry.getValue());
+        }
+        return ret;
+    }
```

Comment:
Isn't mInGroupOverlay valid already? Why not just directly `return mInGroupOverlay`?

Reply:
spotBug does not like it because it point blank sends a private member out, as well as the other half of this function which takes in a hash map as a private member, which would in turn allow modification of the said member from outside the class

while spotBug is okay with `mInGroupOverlay = (HashMap<Integer, Boolean>) cardIdInGroupMap.clone();` and `return (HashMap<Integer, Boolean>)mInGroupOverlay.clone()`, android studio does not like that, seeing it as a problem of `Unchecked cast: 'java.lang.Object' to 'java.util.HashMap<java.lang.Integer,java.lang.Boolean>'`, with a suggested fix of `Try to generify 'ManageGroupCursorAdapter.java'`

if you know a better solution to keep both happy, or you're happy with android studio complaining about the .clone, or any better suggestion, please let me know

Reply:
I care more about spotbugs than Android Studio. Android Studio has gotten stuff wrong in stupid ways before, while spotbugs has prevented serious bugs from slipping into the code base.

Anyway, what about doing this? Seems to make spotBugs happy and seems the simplest code.

```
return new HashMap<>(mInGroupOverlay);
```